services/motion/explore/explore.go

```diff
@@ -0,0 +1,575 @@
+	kb, err := ms.createKinematicBase(ctx, componentName, motionCfg)
+	if err != nil {
+		return false, err
+	}
+
+	// Create motionplan plan
+	planInputs, err := ms.createMotionPlan(ctx, kb, destination.Pose(), worldState, extra)
+	if err != nil {
+		return false, err
+	}
+	var plan motionplan.Plan
+	for _, inputs := range planInputs {
+		input := make(map[string][]referenceframe.Input)
+		input[kb.Name().Name] = inputs
+		plan = append(plan, input)
+	}
+
+	// Start background processes
+	cancelCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Start polling for obstacles
+	ms.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		ms.checkForObstacles(cancelCtx, obstacleDetectors, kb, plan, motionCfg.ObstaclePollingFreqHz)
+	}, ms.backgroundWorkers.Done)
+
+	// Start executing plan
+	ms.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		ms.executePlan(cancelCtx, kb, plan)
+	}, ms.backgroundWorkers.Done)
+
+	for {
+		// this ensures that if the context is cancelled we always return early at the top of the loop
+		if err := ctx.Err(); err != nil {
+			return false, err
+		}
+
+		select {
+		// if context was cancelled by the calling function, error out
+		case <-ctx.Done():
+			return false, ctx.Err()
+
+		// once execution responds: return the result to the caller
+		case resp := <-ms.executionResponseChan:
+			ms.logger.Debugf("execution completed: %s", resp)
+			return resp.success, resp.err
+
+		// if the checkPartialPlan process hit an error return it, otherwise exit
+		case resp := <-ms.obstacleResponseChan:
+			ms.logger.Debugf("obstacle response: %s", resp)
+			if resp.err != nil {
+				return resp.success, resp.err
+			}
+			if resp.success {
+				return resp.success, nil
+			}
+		}
+	}
+}
+
+type moveResponse struct {
+	err     error
+	success bool
+}
+
+// checkForObstacles will continuously monitor the generated transient worldState for obstacles in the given
+// motionplan plan. A response will be sent through the channel if an error occurs, the motionplan plan
+// completes or an obstacle is detected in the given range.
+func (ms *explore) checkForObstacles(
+	ctx context.Context,
+	obstacleDetectors []obstacleDetectorObject,
+	kb kinematicbase.KinematicBase,
+	plan motionplan.Plan,
+	obstaclePollingFrequencyHz float64,
+) {
+	// Constantly check for obstacles in path at desired obstacle polling frequency
+	ticker := time.NewTicker(time.Duration(int(1000/obstaclePollingFrequencyHz)) * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			currentPose := spatialmath.NewZeroPose()
+
+			// Look for new transient obstacles and add to worldState
+			worldState, err := ms.generateTransientWorldState(ctx, obstacleDetectors)
+			if err != nil {
+				ms.obstacleResponseChan <- moveResponse{err: err}
+				return
+			}
+
+			// Check motionplan plan for transient obstacles
+			collisionPose, err := motionplan.CheckPlan(
+				kb.Kinematics(),
+				plan,
```

Comment:
[q]: from conversations offline, I thought that we would only be checking only 10 cm ahead?
where do we augment what the passed in plan is?

Reply:
I have changed it to take in the full plan but only exit the loop when an obstacle is detected a certain distance from the base. This way we do not need to create a partial plan from the full plan and track our location along it, when we dont know what our position is

Reply:
hmmmmmmmm, I believe this implies that we could be in collision with an obstacle that is actually behind the robot, right?

Reply:
Based on my understanding, those obstacles wouldn't be intersecting the path and therefore will not be returned by `CheckPlan`